### PR TITLE
Intly 3289

### PR DESCRIPTION
**Pull request by damienomurchu**
https://issues.jboss.org/browse/INTLY-3793. Add mobile-security-service install to integreatly-operator

### CODE REVIEW

File: pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityservicebackup_types.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
+package v1alpha1
```
Comment (damienomurchu) [inline pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityservicebackup_types.go]: This file can be removed as backups are not being installed in this PR. I will update the PR to remove.

File: pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityserviceapp_types.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package v1alpha1
```
Comment (damienomurchu) [inline pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityserviceapp_types.go]: This file can also be removed I think, as we're not installing an mss app in this PR, just the service and the db used by the service. I will update the PR to remove it.

File: pkg/controller/installation/products/config/manager.go [hunk 3]
```diff
@@ -208,6 +211,14 @@ func (m *Manager) ReadUps() (*Ups, error) {
 	return NewUps(config), nil
 }
 
+func (m *Manager) ReadMobileSecurityService() (*MobileSecurityService, error) {
+	config, err := m.readConfigForProduct(v1alpha1.ProductLauncher)
```
Comment (damienomurchu) [inline pkg/controller/installation/products/config/manager.go]: ProductLauncher incorrect here, I will fix this and update PR

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	"k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	defaultInstallationNamespace = "mobile-security-service"
+	defaultSubscriptionName      = "integreatly-mobile-security-service"
+	clusterName                  = "integreatly-cluster"
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: This is redundant, will remove and update PR

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+func (r *Reconciler) GetPreflightObject(ns string) runtime.Object {
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mobile-security-service",
+			Namespace: ns,
+		},
+	}
+}
+
+// Reconcile reads that state of the cluster for amq streams and makes changes based on the state read
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler.go]:
```suggestion
// Reconcile reads that state of the cluster for amq streams and makes changes based on the state read
// Reconcile reads that state of the cluster for mobile security service and makes changes based on the state read
```

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+// and what is required
+func (r *Reconciler) Reconcile(ctx context.Context, inst *v1alpha1.Installation, product *v1alpha1.InstallationProductStatus, serverClient pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	ns := r.Config.GetNamespace()
+	if ns == "" {
+		// TODO return phase failed as default namespace should be set in NewReconciler
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: Will fix this in PR

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+func (r *Reconciler) handleProgressPhase(ctx context.Context, client pkgclient.Client) (v1alpha1.StatusPhase, error) {
+	r.logger.Debug("checking mobile security service pods are running")
+
+	pods := &v1.PodList{}
+	err := client.List(ctx, &pkgclient.ListOptions{Namespace: r.Config.GetNamespace()}, pods)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, "failed to check mobile security service installation")
+	}
+
+	//4 pods expected - operator, db, service
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: Should be 3, will update comment

File: pkg/controller/installation/products/mobilesecurityservice/reconciler_test.go [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+								Kind:       "MobileSecurityService",
+								APIVersion: mss.SchemeGroupVersion.String(),
+							},
+							ListMeta: metav1.ListMeta{},
+							//ObjectMeta: metav1.ObjectMeta{
+							//	Name: "mss-install-plan",
+							//},
+							//Status: operatorsv1alpha1.InstallPlanStatus{
+							//	Phase: operatorsv1alpha1.InstallPlanPhaseComplete,
+							//},
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler_test.go]: Redundant, will remove in PR

File: pkg/controller/installation/products/mobilesecurityservice/reconciler_test.go [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+								APIVersion: mss.SchemeGroupVersion.String(),
+							},
+							ListMeta: metav1.ListMeta{},
+							//ObjectMeta: metav1.ObjectMeta{
+							//	Name:      "mss-install-plan",
+							//	Namespace: "mobile-security-service",
+							//},
+							//Status: operatorsv1alpha1.InstallPlanStatus{
+							//	Phase: operatorsv1alpha1.InstallPlanPhaseComplete,
+							//},
```
Comment (damienomurchu) [inline pkg/controller/installation/products/mobilesecurityservice/reconciler_test.go]: Redundant, will remove in PR

File: pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityservice_types.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
+package v1alpha1
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityservice_types.go]: Yes, I will check if this is possible

File: pkg/apis/integreatly/v1alpha1/installation_types.go [hunk 11]
```diff
@@ -29,47 +29,50 @@ var (
...
-	VersionLauncher            ProductVersion = "0.1.2"
-	VersionUps                 ProductVersion = "2.3.2"
+	VersionAMQOnline             ProductVersion = "1.2.2"
+	VersionAMQStreams            ProductVersion = "1.1.0"
+	VersionCodeReadyWorkspaces   ProductVersion = "1.2.0.GA"
+	VersionFuseOnOpenshift       ProductVersion = "master"
+	VersionNexus                 ProductVersion = "3.16"
+	VersionLauncher              ProductVersion = "0.1.2"
+	VersionUps                   ProductVersion = "2.3.2"
+	VersionMobileSecurityService ProductVersion = "0.2.2"
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply pkg/apis/integreatly/v1alpha1/installation_types.go]: Yes, was thinking about that. Do you know if we've been able to establish the version dynamically for other products?

[the remark this answers is not in the document]
Comment (philbrookes) [reply pkg/apis/integreatly/v1alpha1/installation_types.go]: yeah, it's finicky, here's a few examples:
https://github.com/integr8ly/integreatly-operator/blob/master/pkg/controller/installation/products/solutionexplorer/reconciler.go#L208
https://github.com/integr8ly/integreatly-operator/blob/master/pkg/controller/installation/products/threescale/reconciler.go#L601
https://github.com/integr8ly/integreatly-operator/blob/master/pkg/controller/installation/products/rhsso/reconciler.go#L208

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+		ConfigManager: configManager,
+		Config:        config,
+		mpm:           mpm,
+		logger:        logger,
+		Reconciler:    resources.NewReconciler(mpm),
+	}, nil
+}
+
+func (r *Reconciler) GetPreflightObject(ns string) runtime.Object {
+	return &appsv1.Deployment{
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: my mistake, this does need to be here, I misread the code.

File: pkg/controller/installation/products/mobilesecurityservice/reconciler.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dbClusterName,
+			Namespace: r.Config.GetNamespace(),
+		},
+		Spec: mobilesecurityservice.MobileSecurityServiceDBSpec{
+			ContainerName:          "database",
+			DatabaseMemoryLimit:    "512Mi",
+			DatabaseMemoryRequest:  "512Mi",
+			DatabaseName:           "mobile_security_service",
+			DatabaseNameParam:      "POSTGRESQL_DATABASE",
```
[the remark this answers is not in the document]
Comment (damienomurchu) [reply pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: In this case I think a hardcoded password is ok, as the DB is not exposed publicly or outside the namespace and is only access by the mss service - wdyt?

[the remark this answers is not in the document]
Comment (damienomurchu) [reply pkg/controller/installation/products/mobilesecurityservice/reconciler.go]: updated with random password now

File: pkg/apis/integreatly/v1alpha1/installation_types.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (damienomurchu) [reply pkg/apis/integreatly/v1alpha1/installation_types.go]: No way to grab this dynamically - the [MSS CR](https://github.com/aerogear/mobile-security-service-operator/blob/master/pkg/apis/mobilesecurityservice/v1alpha1/mobilesecurityservice_types.go) doesn't have any attribute to set this at the moment unfortunately, so will leave it hardcoded at the moment. 
I will look at creating a JIRA for the aerogear team to have a version field added to the MSS

**Pull request merged**